types/jquery/jquery-tests.ts

```diff
@@ -33,63 +32,74 @@ function JQueryStatic() {
         // $ExpectType JQuery<HTMLElement>
         $('span');
 
-        // $ExpectType JQuery<HTMLElement>
-        $('<p></p>');
+        // $ExpectType JQuery<SVGLineElement>
+        $<SVGLineElement>('.mysvgline');
 
-        // $ExpectType JQuery<HTMLElement>
-        $(new HTMLElement());
+        // $ExpectType JQuery<HTMLParagraphElement>
+        $(new HTMLParagraphElement());
 
-        // $ExpectType JQuery<HTMLElement>
-        $([new HTMLElement()]);
+        // $ExpectType JQuery<HTMLParagraphElement>
+        $([new HTMLParagraphElement()]);
 
         // $ExpectType JQuery<{ foo: string; hello: string; }>
         $({ foo: 'bar', hello: 'world' });
 
-        // $ExpectType JQuery<HTMLElement>
-        $($('p'));
+        // $ExpectType JQuery<SVGSVGElement>
+        $($(document.createElementNS("http://www.w3.org/2000/svg", "svg")));
 
         // $ExpectType JQuery<HTMLElement>
         $(function($) {
             // $ExpectType Document
             this;
-            // $ExpectType JQueryStatic<HTMLElement>
+            // $ExpectType JQueryStatic
+            $;
+        });
+
+        // $ExpectType JQuery<Element>
+        $<Element>(function($) {
+            // $ExpectType Document
+            this;
+            // $ExpectType JQueryStatic
             $;
         });
 
         // $ExpectType JQuery<HTMLElement>
         $();
 
+        // $ExpectType JQuery<Element>
+        $<Element>();
+
         // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/19597#issuecomment-378218432
         function issue_19597_378218432() {
-            let myDiv = $(document.createElement('div'));
+            const myDiv = $(document.createElement('div'));
             // $ExpectType JQuery<HTMLDivElement>
             myDiv;
             myDiv.on('click', (evt) => {
-                let target = evt.target;
+                const target = evt.target;
                 // $ExpectType HTMLDivElement
                 target;
             });
-            let myDiv1 = $<HTMLDivElement>(document.createElement('div'));
+            const myDiv1 = $<HTMLDivElement>(document.createElement('div'));
 
-            let myForcedDiv: JQuery<HTMLDivElement> = $(document.createElement('div')) as any;
+            const myForcedDiv: JQuery<HTMLDivElement> = $(document.createElement('div')) as any;
             myForcedDiv.on('click', (evt) => {
-                let target = evt.target; // HTMLDivElement
+                const target = evt.target; // HTMLDivElement
                 // $ExpectType HTMLDivElement
                 target;
             });
-            let myDoc = $(document);
+            const myDoc = $(document);
             // $ExpectType JQuery<Document>
             myDoc;
             myDoc.on('click', (evt) => {
-                let target = evt.target;
+                const target = evt.target;
                 // $ExpectType Document
                 target;
```

Comment:
This is intuitive but wrong!
> The target property can be the element that registered for the event or a descendant of it.

So probably it is HTMLElement or SVGElement.

```ts
myDoc.on('click', (evt) => {
    let target = evt.target;
    // $ExpectType Document|HTMLElement
    target;
});
```

Perhaps the typing should be 
```ts
on(events: string,
        handler: JQuery.EventHandler<TElement | HTMLElement> | JQuery.EventHandlerBase<any, JQuery.Event<TElement>> | false): this;
```

Reply:
This has been brought up before in #17377. There's a couple of solutions to the problem but none of them are particularly clean.

First off, let's keep in mind that there are 4 target properties on `JQuery.Event`.

#### `currentTarget`
The current element within the event bubbling phase.

#### `delegateTarget`
The element that the handler was attached to. With the way the declarations are currently defined, this property is the only one that has the correct type.

#### `relatedTarget`
Another DOM element involved in the event. Set for `mouseout` and `mouseover` events.

#### `target`
The element that initiated the event. This can be the element that registered the event (`delegateTarget`) or a descendant.

---

### Solution 1

The current solution. All target properties have the same generic type. This works well when `TTarget` is some base type (e.g. `HTMLElement` or `SVGElement`), you're only working with 1 DOM, and you don't need to access properties of more specific types. It otherwise produces incorrect types or types that are too general.

### Solution 2a

Set the target properties to a base type. `EventTarget` would be the correct base type but wouldn't be very useful. This leads to correct typing but requires type assertions to be practical.

### Solution 2b

Set the target properties to `any`. This throws away type safety and autocomplete but is the most flexible.

### Solution 3

Introduce type parameters for the other target properties. This allows the user to have the most accurate typings but is a bit unwieldy. This would introduce up to 3 more type parameters on top of the 2 existing. Ordering of the parameters is also significant. They should be arranged to minimize the number of parameters that need to be defined in common cases.

In addition to this, it should also keep in mind additional type parameters that could be introduced to fix the [issue of properties being marked optional on `JQuery.Event` when they are known to exist](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17377).

---

Of all these, I prefer Solution 3 for its accuracy. My concern is that mistakes in the implementation could mean a lot of breaking changes for consumers in order to fix them. I would prefer a more graceful solution.

Reply:
This commit are not compatible with many other typings, with messages like:
`All declarations of 'JQuery' must have identical type parameters.`

Some other error messages:
`Property 'tagName' does not exist on type 'HTMLElement | Text | Comment'.`

Also, the comment style is not playing well with vscode:
```
     * @see \`{@link https://api.jquery.com/find/ }\`
     * @since 1.0
     * @since 1.6
```

Can some of you look at this, and possible fix or revert this pr? @sandersn @leonard-thieu @holgeradam  @Diullei @borislavjivkov  @basarat @johnnyreilly 
If it there are many breaking changes, I would suggest that it should have been a bigger version bump.
I will use the old version until it's resolved.

Reply:
> This commit are not compatible with many other typings, with messages like:
`All declarations of 'JQuery' must have identical type parameters.`

Which typings? All current versions of typings from DefinitelyTyped should be compatible.

> Also, the comment style is not playing well with vscode:

Could you elaborate on what you mean (include a screenshot)?

> If it there are many breaking changes, I would suggest that it should have been a bigger version bump.

`@types/*` packages do not follow semantic versioning. Instead, their major and minor version numbers match the version of the library they represent. I recommend locking `@types/*` packages to a specific version for your needs.

Reply:
I will update all our business typings then.